### ISSUE TITLE
Result of RelatedJoin and SQLRelatedJoin differs

### ISSUE THREAD

**Issue by yudytskiy** [sqlobject/sqlobject#185]
Hi.
I've found different behavour of RelatedJoin and SQLRelatedJoin.
In docs noted:
***RelatedJoin returns a list of results, while SQLRelatedJoin returns a SelectResults object.***
That means that results of RelatedJoin and list(SQLRelatedJoin) must be equal.

May be it depends on my structure, but still I've exepected it should work.

test code:
```
from sqlobject import (sqlhub, connectionForURI, SQLObject, ForeignKey,RelatedJoin, SQLRelatedJoin, DatabaseIndex, StringCol, TimestampCol)
from sqlobject.dberrors import DuplicateEntryError

USER     = 'user'
PASSWORD = 'pass'
HOST     = 'ip'
DB       = 'issue_test'
DRIVER   = 'pymysql'
DATABASE_URI = f'mysql://{USER}:{PASSWORD}@{HOST}/{DB}?driver={DRIVER}'

class Position(SQLObject):
    name      = StringCol(length=128, alternateID=True)
    subs      = SQLRelatedJoin('Position', createRelatedTable=False, intermediateTable='position_sub', joinColumn='position_id', otherColumn='sub_id')
    subs_     = RelatedJoin('Position', createRelatedTable=False, intermediateTable='position_sub', joinColumn='position_id', otherColumn='sub_id')
    maysub    = SQLRelatedJoin('Position', createRelatedTable=False, intermediateTable='position_sub', otherColumn='position_id', joinColumn='sub_id')
    maysub_   = RelatedJoin('Position', createRelatedTable=False, intermediateTable='position_sub', otherColumn='position_id', joinColumn='sub_id')
    last_date = TimestampCol()

class PositionSub(SQLObject):
    position  = ForeignKey('Position', cascade=True)
    sub       = ForeignKey('Position', cascade=True)
    last_date = TimestampCol()
    uniqueIndex = DatabaseIndex(position, sub, unique=True)

def get_or_create(cls, value, field='name'):
    param = {field: value}
    sel = cls.selectBy(**param)
    if sel.count():
        return sel[0]
    return cls(**param)

def create_position_subs(data):
    for pos_name in data.keys():
        try:
            db_pos = get_or_create(Position, pos_name)
            PositionSub(position=db_pos, sub=db_pos)
            for sub_name in data[pos_name]:
                db_pos_sub = get_or_create(Position, sub_name)
                PositionSub(position=db_pos, sub=db_pos_sub)
        except DuplicateEntryError:
            continue


if __name__ == '__main__':
    sqlhub.processConnection = connectionForURI(DATABASE_URI)
    Position.createTable(ifNotExists=True)
    PositionSub.createTable(ifNotExists=True)

    pos_data = {
        "AAA": ["BBB", "CCC"],
        "BBB": ["CCC", "DDD"],
        "DDD": ["CCC"],
        "CCC": ["EEE", "FFF", "GGG"]}

    create_position_subs(pos_data)

    name = 'CCC'
    pos = Position.byName(name)

    print(f'Position {name} substitutes:')
    print('\tRelatedJoin:', [p.name for p in pos.subs_])
    print('\tSQLRelatedJoin:', [p.name for p in pos.subs])

    print(f'Position {name} may be substituted by:')
    print('\tRelatedJoin:', [p.name for p in pos.maysub_])
    print('\tSQLRelatedJoin:', [p.name for p in pos.maysub])
```
the output:
```
Position CCC substitutes:
	RelatedJoin: ['CCC', 'EEE', 'FFF', 'GGG']
	SQLRelatedJoin: ['CCC']
Position CCC may be substituted by:
	RelatedJoin: ['AAA', 'BBB', 'DDD', 'CCC']
	SQLRelatedJoin: ['CCC']
```
Results differ.
SQLRelatedJoin constructs wrong sql statement.

Thanx

**Comment by phdru** [comment]
Hi! Thank you for reporting! I added
```python
    print('\tSQLRelatedJoin:', str(pos.maysub))
```
and the SQL is:
```sql
SELECT position.id, position.name, position.last_date FROM position, position_sub WHERE ((position.id = position_sub.position_id) AND ((position_sub.sub_id = position.id) AND (position.id = 3)))
```
What do you think should be the correct SQL?

**Comment by phdru** [comment]
I only see 2 tables here. Are you sure you want `RelatedJoin`? `RelatedJoin` is intended for 3 tables: two data tables and one intermediate table. See an example at http://sqlobject.org/SQLObject.html#relationships-between-classes-tables

My be you just need `MultipleJoin` for two related tables?

**Comment by yudytskiy** [comment]
Hi!
Thanx for clarifing!
 
The idea is to define some items that may substitute and be substituted by other items.
Let’s say I have an employees with dfferent positions and skills. Some are more skilled and able to substitute less skilled ones.
 
pos_data = {
        "AAA": ["BBB", "CCC"],
        "BBB": ["CCC", "DDD"],
        "DDD": ["CCC"],
        "CCC": ["EEE", "FFF", "GGG"]}
 
employee AAA is more skilled than BBB & CCC
and so on
 
So I’ve come up with idea to use relational structure like the easiest way to achieve this:
a = Position(‘AAA’)
b = Position(‘BBB’)
PositionSub(position=a, sub=b)
…
and we can get all positions that one can beat and all positions that can beat that one
 
so we have just 2 tables:
class Position(SQLObject):
    name      = StringCol(length=128, alternateID=True)
    subs      = SQLRelatedJoin('Position', createRelatedTable=False, intermediateTable='position_sub',  joinColumn ='position_id',  otherColumn ='sub_id')
    maysub    = SQLRelatedJoin('Position', createRelatedTable=False, intermediateTable='position_sub',  otherColumn ='position_id',  joinColumn ='sub_id')
    last_date = TimestampCol()
 
class PositionSub(SQLObject):
    position  = ForeignKey(' Position ', cascade=True)
    sub       = ForeignKey(' Position ', cascade=True)
    last_date = TimestampCol()
    uniqueIndex = DatabaseIndex(position, sub, unique=True)
 
CREATE TABLE `position` (
`id` int ( 11 )  NOT NULL AUTO_INCREMENT ,
`name` varchar ( 128 )  NOT NULL ,
`last_date` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP ,
PRIMARY KEY ( `id` ),
UNIQUE KEY `name` ( `name` )
)  ENGINE = InnoDB AUTO_INCREMENT = 8 DEFAULT CHARSET = utf8mb4 ;
 
CREATE TABLE `position_sub` (
`id` int ( 11 )  NOT NULL AUTO_INCREMENT ,
`position_id` int ( 11 )  DEFAULT NULL ,
`sub_id` int ( 11 )  DEFAULT NULL ,
`last_date` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP ,
PRIMARY KEY ( `id` ),
UNIQUE KEY `uniqueIndex` ( `position_id` , `sub_id` ),
KEY `position_sub_sub_id_exists` ( `sub_id` ),
CONSTRAINT `position_sub_position_id_exists` FOREIGN KEY ( `position_id` )  REFERENCES `position` ( `id` )  ON DELETE CASCADE ,
CONSTRAINT `position_sub_sub_id_exists` FOREIGN KEY ( `sub_id` )  REFERENCES `position` ( `id` )  ON DELETE CASCADE
)  ENGINE = InnoDB AUTO_INCREMENT = 17 DEFAULT CHARSET = utf8mb4 ;
 
so if we take CCC and want to find out which positons CCC may substitute
list(c.subs) must return [EEE, FFF, GGG] less skilled than CCC
an who may substitute CCC
list(c.maysub) must return [AAA, BBB, DDD] more skilled than CCC
 
and it works with RelatedJoin:
Position CCC substitutes:
    RelatedJoin: ['CCC', 'EEE', 'FFF', 'GGG']
    SQLRelatedJoin: ['CCC']
Position CCC may be substituted by:
    RelatedJoin: ['AAA', 'BBB', 'DDD', 'CCC']
    SQLRelatedJoin: ['CCC']
 
as for SQL statement to comply with the logic above it should be something like:
subs:
SELECT position . id ,  position . name ,  position . last_date
FROM position ,  position_sub
WHERE (( position . id =  position_sub . sub_id [ otherColumn ] )  AND ( position_sub . position_id [ joinColumn ] =  3 ))
 
maysub:
SELECT position . id ,  position . name ,  position . last_date
FROM position ,  position_sub
WHERE (( position . id =  position_sub . position_id [ otherColumn ] )  AND ( position_sub . sub_id  [ joinColumn ] =  3 ))
if we look at
subs      = SQLRelatedJoin('Position', createRelatedTable=False, intermediateTable='position_sub',  joinColumn ='position_id',  otherColumn ='sub_id')
maysub    = SQLRelatedJoin('Position', createRelatedTable=False, intermediateTable='position_sub',  otherColumn ='position_id',  joinColumn ='sub_id')
 
thanx!!!
  
>Четверг, 22 июня 2023, 17:32 +03:00 от Oleg Broytman ***@***.***>:
> 
> 
>Hi! Thank you for reporting! I added
>    print('\tSQLRelatedJoin:', str(pos.maysub))
>
>and the SQL is:
>SELECT position.id, position.name, position.last_date FROM position, position_sub WHERE ((position.id = position_sub.position_id) AND ((position_sub.sub_id = position.id) AND (position.id = 3)))
>
>What do you think should be the correct SQL?
>—
>Reply to this email directly,  view it on GitHub , or  unsubscribe .
>You are receiving this because you authored the thread. Message ID:  <sqlobject/sqlobject/issues/185/1602747640 @ github . com>
 
 
--
Игорь Юдыцкий
Отправлено из Почты  Mail.ru

**Comment by phdru** [comment]
This is surely a bug in SO and I hope to return and fix it.

**Comment by phdru** [comment]
I'm returning. Gonna use this simplified test:
```python
#! /usr/bin/env python

from sqlobject import *

__connection__ = "sqlite:/:memory:?debug=1"


class RecursiveGroup(SQLObject):
    name = StringCol(unique=True)
    subgroups = {SQL}RelatedJoin(
        'RecursiveGroup',
        otherColumn='group_id',
        intermediateTable='rec_group_map',
        createRelatedTable=False,
    )


class RecGroupMap(SQLObject):
    recursive_group = ForeignKey('RecursiveGroup')
    group = ForeignKey('RecursiveGroup')


RecursiveGroup.createTable()
RecGroupMap.createTable()

a = RecursiveGroup(name='a')
a1 = RecursiveGroup(name='a1')
a.addRecursiveGroup(a1)
a2 = RecursiveGroup(name='a2')
a.addRecursiveGroup(a2)

print(sorted(a.subgroups, key=lambda x: x.name))
```
With `RelatedJoin` it correctly outputs
```
[<RecursiveGroup 2 name='a1'>, <RecursiveGroup 3 name='a2'>]
```
With `SQLRelatedJoin` the list is empty which is certainly wrong.

**Comment by phdru** [comment]
Can you please test the commit https://github.com/sqlobject/sqlobject/commit/a4ee6cdb15c8e253db0d81cd766396d66bfd2e61 ? It works for me.

**Comment by yudytskiy** [comment]
Hi! Thanx. Works as expected.

**Comment by phdru** [comment]
I released [version 3.10.2](https://pypi.org/project/SQLObject/3.10.2/). Thanks again!
